### PR TITLE
Regularization bug. Tests and fix.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
This PR and the new and clean version of this PR: #8932 

## The bug:
The bug was that when using two models sharing layers in a bigger model, the regularization loss was duplicated. Here are the tests that were written to show the bug.

### test_regularization_shared_layer() -> PASSED
The same layer is used twice in the final model.

### test_regularization_shared_model() -> FAILED
The same model is used twice in the final model

### test_regularization_shared_layer_in_different_models() -> FAILED
Two different model instances are used in the final model. 
But those two model instances refer to a single unique layer.

## Why does it happen?
Because when models/containers gather losses, there is no check on the provenance of the loss. This can be seen in `Layer.add_loss(self, losses, inputs=None)`.

## The fix:
This bug has been fixed. This was done in the function `def losses(self)` as suggested in the previous PR.
We run over the list and de-dupe at this stage. We use the `is` operator. This was the only way I found to compare the origins of the losses.

### CODE REVIEW

**Comment by ozabluda** [comment]
Not commenting on anything other that the quadratic behavior of the uniquification.
1. Is it a problem? Networks can have thousands of losses. Maybe testing with 10,000 is sufficient?
2. If yes, can it be sped up? The following may give you ideas:
https://stackoverflow.com/questions/12730961/removing-duplicates-using-custom-comparisons

**Comment by gabrieldemarmiesse** [comment]
I don't believe the complexity of the fix is an issue. You can do a benchmark though if you think it's going to be a problem. 

I'm more concerned about the use of the `is` operator to compare losses. I wanted to use an unique identifier that I could add as attribute of the tensor in `add_weight`, but I couldn't find a good identifier.

**Comment by ozabluda** [comment]
>I don't believe the complexity of the fix is an issue. You can do a benchmark though if you think it's going to be a problem.

Would be nice to check. N=10,000 losses lead to N^2/2=50,000,000 calls, could be a problem. Nets can have a ~1000 layers, each layer typically has 1-2 regularization losses, BN may have regularization loss(es), then you can have, say, a Siamese triplet of those. Seems that 10K is a reasonable test. A microbenchmark would probably suffice.

>I'm more concerned about the use of the is operator to compare losses. I wanted to use an unique identifier that I could add as attribute of the tensor in add_weight, but I couldn't find a good identifier.

Likewise, but I don't know, so not commenting on this.

File: keras/engine/topology.py [hunk 1]
```diff
@@ -1943,7 +1943,15 @@ def losses(self):
                 losses += layer.get_losses_for(None)
         # Add any potential unconditional model-level loss.
         losses += self.get_losses_for(None)
-        return losses
+
+        losses_without_duplicates = []
```
Comment (fchollet) [inline keras/engine/topology.py]: - Wouldn't `list(set(losses))` do the same thing more efficiently?
- Wouldn't testing for object identity (`is`) be an issue for losses that are integers or floats? (exact same behavior as use of `id`).

Comment (gabrieldemarmiesse) [reply keras/engine/topology.py]: I agree with you. How do you suggest that we test for object identity? I could make a pseudo static variable in add_weight, but I'm not sure it's cleaner than using is or id.

**Comment by gabrieldemarmiesse** [comment]
- So, when creating the weights, I can give the regularization tensor a unique id `_weight_id` (using a "static" variable). That's what I use later. 
- Keras doesn't change its behavior for losses that don't have `_weight_id`.
- I also change the way the deduping is done. It should be a bit faster now for big networks.

**Comment by ozabluda** [comment]
>Keras doesn't change its behavior for losses that don't have _weight_id.

1. will it work for saved/loaded models?
2. will it work for models saved before this PR, loaded after this PR (and then used Siamese-style)?

**Comment by ozabluda** [comment]
FWIW, the (now fixed) quadratic behavior seems to be a problem. Running the microbenchmark below 
takes 3.5 sec with 10,000, 12.3 sec with 20,000 (demonstrating the quadratic behavior):
```
losses = []
losses_without_duplicates = []
for i in range(10000):
    def f():
        pass
    losses.append(f)

for i, loss in enumerate(losses):
    for loss2 in losses[i + 1:]:
        if loss is loss2:
            break
    else:
        losses_without_duplicates.append(loss)
```

**Comment by ozabluda** [comment]
The now fixed code is much faster.  Running the microbenchmark below takes 0.12 sec with 10,000, 0.34 sec with 20,000.
```
losses = []
losses_without_duplicates = []
for i in range(20000):
    class f:
        pass
    losses.append(f())

tmp_dic = {}
for loss in losses:
    if hasattr(loss, "_weight_id"):
        tmp_dic[loss._weight_id] = loss
    else:
        losses_without_duplicates.append(loss)
 
    losses_without_duplicates += list(tmp_dic.values())
```

File: keras/engine/topology.py [hunk 2]
```diff
@@ -413,7 +415,10 @@ def add_weight(self,
                             name=name,
                             constraint=constraint)
         if regularizer is not None:
-            self.add_loss(regularizer(weight))
+            reg_tensor = regularizer(weight)
+            reg_tensor._weight_id = Layer._weight_id
+            Layer._weight_id += 1
```
Comment (fchollet) [inline keras/engine/topology.py]: This is not thread-safe.

File: keras/engine/topology.py (hunk 1 above)
Comment (fchollet) [reply keras/engine/topology.py]: We could do:

```python
unique_tensors = list(set([x for x in losses if not isinstance(x, (float, int))]))
non_tensors = [x for x in losses if isinstance(x, (float, int))]
losses = unique_tensors + non_tensors
```

What do you think?

File: keras/engine/topology.py (hunk 2 above)
Comment (gabrieldemarmiesse) [reply keras/engine/topology.py]: Ok, I completely misunderstantood what you meant earlier with the use of id(), my bad. I'll change it and use set().

**Comment by gabrieldemarmiesse** [comment]
in TestBackend.test_random_normal with CNTK, I got an assertion error. Re-trigerring the tests.
assert 0.016304660588502884 < 0.015 was just bad luck I guess.

**Comment by fchollet** [comment]
We may have to change a random seed in the tests to fix it.

**Comment by gabrieldemarmiesse** [comment]
Should I do it in this PR or open another one?

**Comment by fchollet** [comment]
This PR, so the tests can pass.

**Comment by gabrieldemarmiesse** [comment]
Thanks!

**Pull request merged**